### PR TITLE
feat(codec): molecule friendly error message

### LINKED ISSUE #396: [feature request] validator for codec when packing the object
## Motivation

When I call the `codec.pack`, especially when packing a large or nested object, i'm hard to troubleshoot which part of the object is incorrect, and it would be much easier to debug if the pack could tell me exactly which part is incorrect

## How

```ts
const RGB = array(Uint8, 3)

//       overflow
//         👇
RGB.pack([0xfff, 0xff, 0xff]) // [0] is incorrect, expected ..., actual ...
```

```ts
const RGB = array(Uint8, 3)
const Styled = struct({
  background: RGB,
  borderColor: RGB
}, ['background', 'borderColor']);

Styled.pack({
  //          overflow
  //             👇
  background: [0xfff, 0xff, 0x00],
  borderColor: [0x12, 0x34, 0x56],
}) // error: background.[0] is incorrect, expected ..., actual ...

```

### `.validate` for Codec

Can we add a `validatePack` method to the codec to replace the validators in toolkit

```ts
WitnessArgs.validatePack({...}) // throw an error if the param is invalid
```

https://github.com/nervosnetwork/lumos/blob/d5e7636736e22c3247d36b4160213e74ed23dd0e/packages/toolkit/src/validators.js#L378-L398

### Unpack

For unpack, is there a better way to tell me what went wrong when unpack goes wrong?

### PR DESCRIPTION
**Pull request by IronLu233**
https://github.com/nervosnetwork/lumos/issues/396

### CODE REVIEW

File: packages/codec/tests/molecule.test.ts [hunk 1]
```diff
@@ -214,3 +215,225 @@ test("test fixed hex bytes", (t) => {
...
 });
+test("test simple array codec error", (t) => {
+  const codec = array(Uint8, 3);
+  try {
+    codec.pack([0x1, 0xfff, 3]);
+    t.fail();
+  } catch (e) {
+    t.truthy(
+      (e as CodecExecuteError).message.includes(
+        "Expect type Uint8BE at input[1] but got error: Value must be between 0 and 255, but got"
```
Comment (zhangyouxin) [inline packages/codec/tests/molecule.test.ts]: Uint8 only contains 1 byte, so BE and LE is the same for it. Can you add some other test cases for Uints, like `Uint32LE`

File: packages/codec/src/base.ts [hunk 2]
```diff
@@ -114,3 +114,47 @@ export function createFixedBytesCodec<Unpacked, Packable = Unpacked>(
     }),
   };
 }
+
+export const CODEC_OPTIONAL_PATH = Symbol("?");
```
Comment (zhangyouxin) [inline packages/codec/src/base.ts]: better create an `error.ts`

**Comment by PainterPuppets** [comment]
Can you add description and purpose about this pull request?

File: packages/codec/tests/molecule.test.ts [hunk 3]
```diff
@@ -214,3 +215,225 @@ test("test fixed hex bytes", (t) => {
...
+    t.fail();
+  } catch (e) {
+    t.truthy(
+      (e as CodecExecuteError).message.includes(
+        "Expect type Uint8BE at input.f2 but got error: Value must be between 0 and 255, but got"
+      )
+    );
+  }
+
+  try {
```
Comment (PainterPuppets) [inline packages/codec/tests/molecule.test.ts]: maybe use `const error = t.throws(...)`, you can search for `t.throws` to find many usage example

Comment (IronLu233) [reply packages/codec/tests/molecule.test.ts]: https://github.com/avajs/ava-docs/blob/main/zh_CN/readme.md#throwsfunctionpromise-error-message

A good suggestion. we can test the error message in the second parameter.

**Review by PainterPuppets — COMMENTED** [review]
I'm wondering if it's possible to use the higher-order function to do non-intrusive modifications instead of directly changing the `codec`?

maybe example like that
```
withReadableError(array)(xxxx)
```

File: packages/codec/tests/molecule.test.ts (hunk 1 above)
Comment (IronLu233) [reply packages/codec/tests/molecule.test.ts]: OK. I also add some cases for bigger units.

**Comment by IronLu233** [comment]
> I'm wondering if it's possible to use the higher-order function to do non-intrusive modifications instead of directly changing the `codec`?
> 
> maybe example like that
> 
> ```
> withReadableError(array)(xxxx)
> ```

**It is difficult.**

We want track the property access path in codec pack input, but these key only appears in `Array.prototype.map/reduce`
methods.

When code run out of these high order function, we can not get the **current** key anymore. so I think we have to modify some code in codec pack logic.

**Comment by homura** [comment]
> When code run out of these high order function, we can not get the current key anymore. so I think we have to modify some code in codec pack logic.

vote for this. but i want to know it is possible to refactor `createObjectCodec` and `createArrayCodec` to provide the friendly message? because it's not just molecule that can be affected by nested error. for struct/table/array/vector will call the `createObject|ArrayCodec` to report the friendly message

**Comment by IronLu233** [comment]
> > When code run out of these high order function, we can not get the current key anymore. so I think we have to modify some code in codec pack logic.
> 
> vote for this. but i want to know it is possible to refactor `createObjectCodec` and `createArrayCodec` to provide the friendly message? because it's not just molecule that can be affected by nested error. for struct/table/array/vector will call the `createObject|ArrayCodec` to report the friendly message

Maybe ES6 proxy class can implement this feature.

**Comment by homura** [comment]
> Maybe ES6 proxy class can implement this feature.

it can be simpler

```ts
function struct(shape, fields) {
  const codec = createObjectCodec(...)

  return {
    pack(packable) {
      const packedFields = fields.map(field => codec[field].pack(packable[item]))
      return concat(...packedFields) 
    }
    unpack() {
      // ...
    }
  }
}
```

**Comment by IronLu233** [comment]
> > Maybe ES6 proxy class can implement this feature.
> 
> it can be simpler
> 
> ```ts
> function struct(shape, fields) {
>   const codec = createObjectCodec(...)
> 
>   return {
>     pack(packable) {
>       const packedFields = fields.map(field => codec[field].pack(packable[item]))
>       return concat(...packedFields) 
>     }
>     unpack() {
>       // ...
>     }
>   }
> }
> ```

I found a way to track layout struct(`array`, `table`) error without modify it's codec implementations.

This can be in another PR in https://github.com/nervosnetwork/lumos/pull/408

File: packages/codec/src/error.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
+export const CODEC_OPTIONAL_PATH = "______?______";
```
Comment (homura) [inline packages/codec/src/error.ts]:
```suggestion
export const CODEC_OPTIONAL_PATH = "__molecule_option__";
```

Comment (homura) [inline packages/codec/src/error.ts]:
```suggestion
// lc for lumos codec
export const CODEC_OPTIONAL_PATH = "__lc_option__";
```
i think the `____?____` is too magic when debugging

File: packages/codec/src/error.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  constructor(message: string, public expectedType: string) {
+    super(message);
+  }
+}
+
+const CODEC_EXECUTE_ERROR_NAME = "CodecExecuteError";
+export function isCodecExecuteError(
+  error: unknown
+): error is CodecExecuteError {
+  return (error as Error)?.name === CODEC_EXECUTE_ERROR_NAME;
```
Comment (homura) [inline packages/codec/src/error.ts]:
```suggestion
  if (!(error instanceof Error)) return false;
  return error.name === CODEC_EXECUTE_ERROR_NAME;
```

we should avoid using `as` if possible

Comment (IronLu233) [reply packages/codec/src/error.ts]: I agree. Although at here I use optional chaining for preventing access nullable object.

**Comment by IronLu233** [comment]
Seems everything is OK. Can you please help review again? @homura

**Review by homura — APPROVED** [review]

**Pull request merged**